Adds @listen and @gestureEventListener decorators.

aarondrabeck opened the pull request:
Issue https://github.com/Polymer/polymer-decorators/issues/14

Don't be shy if you see something that could be done better, I am happy to correct it.  Thanks!

aomarks commented:
Thanks for the PR! Probably won't be able to send feedback until next week.

justinfagnani commented:
@aarondrabeck I think it'd be nicer to only have one `@listen` decorator that ends up calling [`this._addEventListenerToNode`](1), which is overridden by `GestureEventListeners` to handle gestures by default.

As for storing the listeners on the proto, this is something that `Polymer.Class` has to support declarative listeners. It'd be good to reuse a similar mechanism if possible. Rather than patching up the prototype and manually implementing a super call with `ready.apply(this, args);` I think it'd be good to just require that the base class implement declarative listeners and vend a declarative listeners mixin. See [the override of `_applyListeners`](https://github.com/Polymer/polymer/blob/master/lib/legacy/class.html#L194).

Also, for performance it may be better to default to adding these listeners async.

@sorvell and @azakus may have opinions on vending a declarative listener mixin, which I believe was removed before 2.0. Would it be better if it were async?

[1]: `https://github.com/Polymer/polymer/blob/6af12ac1afb920df17c4f1e30736b15184f14a8c/lib/mixins/template-stamp.html#L466`

aarondrabeck commented:
@justinfagnani  This would make these decorators much more simple which would be fantastic.  

What if a consumer wanted to mix both GestureEventListeners and DeclarativeListeners in the same class?  Add a type to the @listen('', 'tap', 'gesture') decorator?  

Currently I do not think you can call _addEventListenerToNode prior to ready because the target node is not available at the time the decorator is executed (unless the target is window or this is suppose). It would have to work more like the new declarative mixin and delay registration, correct?

justinfagnani commented:
I think you have it right. The declarative listeners mixin would override `ready()` and call `_addEventListenerToNode` for each stored listener. The `@listen` decorator would just store the decorated methods as listeners.

This all depends on what @sorvell thinks about a declarative listener mixin. We could add one to this repository too.

aarondrabeck commented:
Awesome. 

If I find some thing this weekend, I will create a declarative listener mixin for this repo in case it never makes in to Polymer Core.  Then I can adjust the decorators as discussed.

aarondrabeck commented:
@justinfagnani @aomarks I have updated the pull request per your feedback above.  

Until @sorvell and @azakus decide if you will vend a declarative listener mixin, I went ahead and added one to this repo.  

Now the @listen decorator can be imperative or gesture depending only on which mixin's you apply...which is really convienent.

The only thing left is to decide how to make these async.  Currently the [_addEventListenerToNode](https://github.com/Polymer/polymer/blob/master/lib/mixins/template-stamp.html#L466) method doesn't expose a way to set options.

aomarks commented:
Hi @aarondrabeck, sorry for the long delay. We haven't forgotten about this PR, I'll try to get some movement this week.

aarondrabeck commented:
Thanks!  Let me know if you would like to see any changes made.

justinfagnani commented on src/decorators.ts:
```diff
@@ -145,3 +145,26 @@ function _query(
+    }
+
+    proto.constructor.listeners.push(
+        {node: target, handler: methodName, eventName});
```
Rather than storing the method name, look up the method here. You'll avoid property lookups during `ready()`, and be able to give earlier warnings if the decorator isn't applied to a method.

justinfagnani commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,67 @@
+            let node = typeof listener.node === 'string' ? this.$[listener.node] : listener.node;
+
+            if (typeof method !== 'function') {
+              console.warn('Failed to register ' + listener.eventName + ' listener on ' + listener.handler +
```
I think this check should be done in the decorator.

justinfagnani commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,67 @@
+              console.warn('Failed to register ' + listener.eventName + ' listener on ' + listener.handler +
+                ' handler.');
+            } else {
+              this._addEventListenerToNode(node, listener.eventName, (...args) => {
```
This can be slightly simpler:

```js
this._addEventListenerToNode(node, listener.eventName, method.bind(this));
```

justinfagnani commented on README.md:
```diff
@@ -145,6 +165,17 @@ class TestElement extends Polymer.Element {
+  //
+  // ex. class TestElement extends Polymer.DeclarativeEventListeners(Polymer.GestureEventListeners(Polymer.Element))
+  //
+  @listen('tap', 'submitButton')
```
describe the arguments in the comment above

justinfagnani commented on src/decorators.ts:
```diff
@@ -145,3 +145,26 @@ function _query(
+ * to.
+ *
+ */
+export function listen(eventName: string, target: string|Window|Document) {
```
I think the type for `target` should probably be just be `string|EventTarget`. It might be weird to specify a Node other than `window` or `document`, but it's possible.

justinfagnani replied:
I try to declare all exports as `const` because inter-module bindings are faster for them. Arrow functions should be faster too, so:

```ts
export const listen = (eventName: string, target: string|EventTarget) =>
    (proto: any, methodName: string) => {
      // ...
    };
```

justinfagnani commented on src/decorators.ts:
```diff
@@ -145,3 +145,26 @@ function _query(
+ *
+ */
+export function listen(eventName: string, target: string|Window|Document) {
+  return (proto: any, methodName: string): any => {
```
remove the return type, since this function doesn't return a value.

justinfagnani commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,67 @@
+       * @mixinClass
+       * @implements {Polymer_DeclarativeEventListeners}
+       */
+      class DeclarativeEventListeners extends superClass {
```
Declare a static listeners property.

justinfagnani commented on test/integration/decorators.html:
```diff
@@ -32,7 +34,19 @@
 
+    <test-fixture id="imperative-listener-test-element-fixture">
+      <template>
+        <imperative-listener-test-element></imperative-listener-test-element>
```
imperative -> declarative?

aomarks commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,67 @@
+          super.ready();
+
+          if (!this.constructor.listeners)
+            return;
```
Formatting yes. Run `npm run format`. No linter.

aarondrabeck commented on README.md:
```diff
@@ -145,6 +165,17 @@ class TestElement extends Polymer.Element {
     console.log(`foo is now: ${newFoo}, bar is now ${newBar}`);
   }
+  
+  // @listen can register both imperative or gesture listeners.  Your class must extend the
```
You are correct, this can be better stated:

 ` @listen can register both gesture and non-gesture listeners declaratively.  To accomplish this, you must extend from both Polymer.GestureEventListeners and Polymer.DeclarativeEventListeners. 
`

aarondrabeck replied:
Take 3:   

````
  // @listen registers event listeners declaratively. 
  //
  // To use this decorator, your element class must extend from the supplied Polymer.DeclarativeEventListeners mixin.
  // Ex. class TestElement extends Polymer.DeclarativeEventListeners(Polymer.Element)
  //
  // The @listen decorator can also declaratively register gesture events. To accomplish this, 
  // you must extend from both Polymer.GestureEventListeners and Polymer.DeclarativeEventListeners. 
  //
  // Ex. class TestElement extends Polymer.DeclarativeEventListeners(Polymer.GestureEventListeners(Polymer.Element))
  //
  // listen(eventName: string, target: string|EventTarget)
  // @param eventName A string representing the event type to listen for
  // @param target A single element by id or EventTarget to target
  //

````

aomarks commented:
Ping @aarondrabeck?

aarondrabeck commented:
Sorry I have most of this addressed on a branch then the holidays
happened.  Currently traveling but I plan to finish it up as soon as
possible.  Sorry about the delay.

On Fri, Dec 22, 2017, 4:04 PM Alexander Marks <notifications@github.com>
wrote:

> Ping @aarondrabeck <https://github.com/aarondrabeck>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Polymer/polymer-decorators/pull/15#issuecomment-353693103>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJXrAToDK0O0MAnUC1ZO2aF7UROPtcd2ks5tDEOegaJpZM4PoB4d>
> .
>
-- 
Regards,

Aaron D

aomarks commented:
> Sorry I have most of this addressed on a branch then the holidays
> happened.  Currently traveling but I plan to finish it up as soon as
> possible.  Sorry about the delay.

Great, no problem at all, we're mostly on holiday now too.

aarondrabeck commented:
@justinfagnani Thanks for the comments.  I have updated this PR to address them.  Let me know if any further suggestions or improvments.  

cc @aomarks

Happy Holidays

aomarks commented on README.md:
````diff
@@ -93,15 +93,35 @@ To enable Metadata Reflection:
   <script src="/bower_components/reflect-metadata/Reflect.js"></script>
   ```
+  
+To enable the `@listen` decorator:
````
Maybe add an Event Listeners heading, so it's not under the Metadata Reflection heading.

aomarks commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,79 @@
+     */
+    Polymer.DeclarativeEventListeners = Polymer.dedupingMixin(superClass => {
+
+      /**
```
Extra indent space, and in a few places below too.

aomarks commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,79 @@
+      /**
+     * @polymer
+     * @mixinClass
+     * @implements {Polymer_DeclarativeEventListeners}
```
These interfaces for JS Compiler are generated by https://github.com/Polymer/gen-closure-declarations. Maybe we should be running that and including an `externs.js` here. Same for TypeScript via https://github.com/Polymer/gen-typescript-declarations.

Don't worry about it for now though, I can hook that up after this PR is in.

aomarks commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,79 @@
+        /**
+        * Adds to stored listeners to subscribe to on ready() 
+        *
+        * @param {string|Object} target Element/node for listener to target.
```
`!Object`

aomarks commented on mixins/declarative-event-listeners.html:
```diff
@@ -0,0 +1,79 @@
+        *
+        * @param {string|Object} target Element/node for listener to target.
+        * @param {string} eventName Name of event to listen for.
+        * @param {fucntion} handler Function which receives the notification.
```
`function`

aomarks commented on src/decorators.ts:
```diff
@@ -145,3 +145,20 @@ function _query(
 }
+
+/**
+ * A TypeScript property decorator factory that causes the decorated method to
```
Document that this requires the mixin be applied.

aomarks commented on test/integration/elements/declarative-listener-test-element.ts:
```diff
@@ -0,0 +1,40 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
```
2017, and in all other new source files

aomarks commented on src/decorators.ts:
```diff
@@ -145,3 +145,20 @@ function _query(
+ *
+ * @param eventName A string representing the event type to listen for
+ * @param target A single element by id or EventTarget to target
+ *
```
Extra line.

aomarks commented on test/integration/polymer.d.ts:
```diff
@@ -2,6 +2,16 @@
 declare namespace Polymer {
   class Element extends HTMLElement {
-    ready():void;
+    ready(): void;
```
FYI generated types are in master now: https://github.com/Polymer/polymer/blob/master/types. Not sure how long it will take to get into a release though, so happy to just leave these as-is and switch over to the generated types later.

aarondrabeck replied:
Oh very nice!

aarondrabeck commented:
@aomarks updated PR per comments.  Thanks!
